torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -119,9 +119,12 @@ c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
   if (hasValue) {
     c10::intrusive_ptr<JitFuture> jitFuture =
         c10::make_intrusive<JitFuture>(PyObjectType::get());
-
+    // Save and pass thread local state into the callback
+    at::ThreadLocalState tls_state;
```

Comment:
Is it reasonable to have a `addCallbackWithTLSState` that will abstract over passing in the state like this? I had to do this quite a few times in order to get the profiler state propagated everywhere necessary.

Reply:
How expensive is this TLS? Is it possible to just do this when profiler is enabled?

Reply:
In the constructor it saves some state such as dispatch key, debug info (such as profiler info), thread local callbacks, etc - so there are a couple of writes, but no computationally expensive work going on. The alternative I guess is to add a check for the profiler being enabled, although, the profiler is usually not enabled, so this would also add extra overhead to every `addCallback`. Maybe @ilia-cher has some insight too.

Reply:
the most comp. intensive part in creating/copying ThreadLocalState is copying of vector of std::function; when profiler is not set this vector is empty; the rest in ThreadLocalState are just bools/ints and TLS reads/writes

Reply:
Per @ilia-cher's comment and a discussion we had offline, we concluded it's fine to propagate TLS since it's only significantly more expensive when profiler is enabled (copying vector of callbacks), which is to be expected as part of the profiling overhead.

Reply:
That was the conclusion, though I can add that and refactor this in a follow up PR.